ibmcloud/image/build.sh

```diff
@@ -206,69 +257,163 @@ umount "$src_mnt/dev"
+    echo "PARTUUID=$BOOT_UUID    /boot-se    ext4  norecovery 1 2" >> ${dst_mnt}/etc/fstab
+    echo "adding luks keyfile for fs"
+    cp rootkey.bin ${dst_mnt}/etc/keys/luks-$(blkid -s UUID -o value /dev/mapper/$LUKS_NAME).key
+    chmod 600 ${dst_mnt}/etc/keys/luks-$(blkid -s UUID -o value /dev/mapper/$LUKS_NAME).key
+    echo "#This file was auto-generated" >  ${dst_mnt}/etc/crypttab
+    echo "$LUKS_NAME UUID=$(blkid -s UUID -o value ${dst_part}) /etc/keys/luks-$(blkid -s UUID -o value /dev/mapper/$LUKS_NAME).key luks,discard,initramfs" >> ${dst_mnt}/etc/crypttab
+    chmod 744 ${dst_mnt}/etc/crypttab
+    # Disable virtio_rng
+    echo "#don't trust rng from hypervisor"  > ${dst_mnt}/etc/modprobe.d/blacklist-virtio.conf
+    echo "blacklist virtio_rng"             >> ${dst_mnt}/etc/modprobe.d/blacklist-virtio.conf
+
+    # Favor loading of TRNG module for newer Z machines
+    echo 's390_trng' >> ${dst_mnt}/etc/modules
+
+    # Prep files needed for mkinitrd (if running encrypted /)
+    echo "KEYFILE_PATTERN=\"/etc/keys/*.key\"" >> ${dst_mnt}/etc/cryptsetup-initramfs/conf-hook
+    echo "UMASK=0077" >> ${dst_mnt}/etc/initramfs-tools/initramfs.conf
+    cat <<END > ${dst_mnt}/etc/zipl.conf
+[defaultboot]
+default=linux
+target=/boot-se
+
+targetbase=/dev/vda
+targettype=scsi
+targetblocksize=512
+targetoffset=2048
+
+[linux]
+image = /boot-se/se.img
+END
+    echo "updating initial ram disk"
+    chroot "$dst_mnt" update-initramfs -u || true
+    echo "!!! Bootloader install errors prior to this line are intentional !!!!!" 1>&2
+    printf "\nGenerating an IBM Secure Execution image\n"
+    # Clean up kernel names and make sure they are where we expect them
+    KERNEL_FILE=$(readlink ${dst_mnt}/boot/vmlinuz)
+    echo "using kernel: ${KERNEL_FILE}"
+    INITRD_FILE=$(readlink ${dst_mnt}/boot/initrd.img)
+    echo "using initrd: ${INITRD_FILE}"
+    SE_PARMLINE="root=/dev/mapper/$LUKS_NAME console=ttysclp0 quiet panic=0 rd.shell=0 blacklist=virtio_rng swiotlb=262144"
+    echo "${SE_PARMLINE}" > ${dst_mnt}/boot/parmfile
+    host_keys=""
+    for i in $(ls ${HOST_KEYS_DIR}/*.crt); do
+        echo "found host key file: \"${i}\""
+        host_keys+="-k ${i} "
+    done
+    /usr/bin/genprotimg \
+        -i ${dst_mnt}/boot/${KERNEL_FILE} \
+        -r ${dst_mnt}/boot/${INITRD_FILE} \
+        -p ${dst_mnt}/boot/parmfile \
+        --no-verify \
+        ${host_keys} \
+        -o "$dst_mnt"/boot-se/se.img
+    # exit and throw an error if no se image was created
+    [ ! -e $dst_mnt/boot-se/se.img ] && exit 1
+    # if building the image succeeded wipe /boot
+    rm -rf ${dst_mnt}/boot/*
+    printf "\nRunning zipl to prepare boot partition\n"
+    chroot $dst_mnt zipl --targetbase $tmp_nbd \
+    --targettype scsi \
+    --targetblocksize 512 \
+    --targetoffset 2048 \
+    --target /boot-se \
+    --image /boot-se/se.img
+    printf "\nClean luks keyfile\n"
+    shred -u rootkey.bin
```

Comment:
If `rootkey.bin` is a very secret one,  how about using a temporary file on a tmpfs, instead of a regular file on the current directory?

A more secure way is to use a temporary variable, but doing it in bash looks complicated.
https://unix.stackexchange.com/a/278644

Reply:
updated, use tmpfs here.